src/constraints.jl

```diff
@@ -0,0 +1,738 @@
+    return JuMP.add_constraint(model, _DisjunctConstraint(obj, tag.indicator), name)
+end
+
+# General fallback for additional arguments
+function _disjunction(
+    _error::Function,
+    model::JuMP.Model, # TODO: generalize to AbstractModel
+    structure,
+    name::String,
+    extra...
+)
+    for arg in extra
+        _error("Unrecognized argument `$arg`.")
+    end
+end
+
+"""
+    disjunction(
+        model::JuMP.Model, 
+        structure, 
+        name::String = ""; 
+        nested::Bool = false,
+        indicator::Union{LogicalVariableRef, Nothing} = nothing
+    )
+
+Function to add a [`Disjunction`](@ref) to a [`GDPModel`](@ref). 
+If no indicator is passed, an anonymous [`LogicalVariable`](@ref) is created.
+"""
+function disjunction(
+    model::JuMP.Model, 
+    structure, 
+    name::String = ""; 
+    nested::Bool = false,
+    indicator::Union{LogicalVariableRef, Nothing} = nothing
+) # TODO add kw argument to build exactly 1 constraint
+    if !isnothing(indicator)
+        return _disjunction(error, model, structure, name, DisjunctConstraint(indicator))
+    elseif nested
+        return _disjunction(error, model, structure, name, DisjunctConstraint)
+    else
+        return _disjunction(error, model, structure, name)
+    end
+end
+
+################################################################################
+#                              LOGICAL CONSTRAINTS
+################################################################################
+# _first_order_ops = (
+#     (:Ξ, :exactly),
+#     (:Λ, :atmost), 
+#     (:Γ, :atleast)
+# )
+# for (ops, set) in zip(_first_order_ops, [Exactly, AtMost, AtLeast])
+#     for op in ops
+#         function JuMP.parse_constraint_call(_error::Function, ::Bool, ::Val{op}, val, lvec)
+#             build_code = :(JuMP.build_constraint($(_error), $(esc(lvec)), $set($(esc(val)))))
+#             return :(), build_code
+#         end
+#     end
+# end
+"""
+    function JuMP.parse_constraint_call(
+        _error::Function, 
+        is_vectorized::Bool, 
+        ::Val{:⇒}, 
+        lhs, 
+        rhs
+    )
+
+Extend `JuMP.parse_constraint_call` to avoid order of precedence when 
+using the logical operator ⇒ when building a 
+logical proposition constraint.
+"""
+function JuMP.parse_constraint_call(
+    _error::Function, 
+    is_vectorized::Bool, 
+    ::Val{:⇒}, 
+    lhs, 
+    rhs
+)
+    op = :⇒
+    error_msg = 
+        "The MOI set in the logical constraint $lhs $op $rhs was not identified. " *
+        "If using the `func in set` notation, the set `MOI.EqualTo{Bool}(value)` " *
+        "should be preceded by `in` or `∈`. Otherwise `func == Bool` should be used " *
+        "to define the constraint."
+    rhs0, set0 = rhs.args[2:3]
+    if rhs.args[1] in (:in, :∈) && set0.head == :call && set0.args[2] isa Bool    
+        set = set0
+    elseif rhs.args[1] == :(==) && set0 isa Bool
+        set = _MOI.EqualTo(set0)
+    else
+        _error(error_msg)
+    end
```

Comment:
The proper way to do this would be to just give the raw `set0` to a function in the build code that will properly check it at runtime.

Reply:
So you would have something like `code = :( build_constraint(_error, func, _process_raw_set(_error, set0)) )` where `_process_raw_set` is a collection of overloaded functions to deal with the set via dispatch as you want to.

Reply:
I see that `<-->` is already built into JuMP, so we cannot touch that anyways.

Reply:
And if we do put just an error, a safer place to do so would be with `operator_to_set`.

Reply:
The `parse_code` is simply executed immediately before the `build_code`, but again this detail becomes uncesessary if we choose not to extend `parse_constraint_call`.

In JuMP, if we use `<--->` this happens:
```julia
@constraint(m, y1 <--> y2 == true)
ERROR: LoadError: At REPL[12]:1: `@constraint(m, y1 <--> y2 == true)`: Invalid right-hand side `y2 == true` of reified constraint. Expected constraint surrounded by `{` and `}`.
```
Thus, I would suggest we forego the use of both `-->` and `<-->` to avoid any confusion.

I saw the recursive strategy, we could explore something like this, but again there are a lot of edge cases to consider. And accounting for vectorized syntax compounds the complexity further. I think the problem is that we are working against the intended design of `parse_constraint_call` which is intended to enable constraints of the form `@constraint(m, myfunc(x, y))`, not intercept specifics of the constraint LHS to change the order of operations.

For this, my opinion would be to not extend `parse_constraint_call` and leave the default error message for now. If people complain, we could then extend `operator_in_set` for `⇔` and `⇒` to throw a better error. The consequence to this is that is would prevent other extensions from creating constraints using these operators to describe some new set type.